apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/chat/ChatSections.kt

```diff
@@ -0,0 +1,254 @@
+package chat.simplex.common.views.chat
+
+import androidx.compose.runtime.toMutableStateList
+import chat.simplex.common.model.*
+import chat.simplex.common.platform.chatController
+import chat.simplex.common.platform.chatModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import kotlin.math.max
+
+const val MAX_SECTION_SIZE = 500
+
+enum class ChatSectionArea {
+  Bottom,
+  Current,
+  Destination
+}
+
+data class ChatSectionAreaBoundary (
+  var minIndex: Int,
+  var maxIndex: Int,
+  val area: ChatSectionArea
+)
+
+data class ChatSection (
+  val items: MutableList<SectionItems>,
+  val boundary: ChatSectionAreaBoundary,
+  val itemPositions: MutableMap<Long, Int>
+)
+
+data class SectionItems (
+  val mergeCategory: CIMergeCategory?,
+  val items: MutableList<ChatItem>,
+  val revealed: Boolean,
+  val showAvatar: MutableSet<Long>,
+)
+
+data class ChatSectionLoader (
+  val position: Int,
+  val sectionArea: ChatSectionArea
+) {
+  fun prepareItems(items: List<ChatItem>): List<ChatItem> {
+    val chatItemsSectionArea = chatModel.chatItemsSectionArea
+    val itemsToAdd = mutableListOf<ChatItem>()
+    for (cItem in items) {
+      val itemSectionArea = chatItemsSectionArea[cItem.id]
+      if (itemSectionArea == null) {
+        itemsToAdd.add(cItem)
+      } else if (itemSectionArea != this.sectionArea) {
+        val targetSection = when (itemSectionArea) {
+          ChatSectionArea.Bottom -> ChatSectionArea.Bottom
+          ChatSectionArea.Current -> if (this.sectionArea == ChatSectionArea.Bottom) ChatSectionArea.Bottom else ChatSectionArea.Current
+          ChatSectionArea.Destination -> if (this.sectionArea == ChatSectionArea.Bottom) ChatSectionArea.Bottom else ChatSectionArea.Destination
+        }
+
+        chatItemsSectionArea.filter { it.value == itemSectionArea }
+          .forEach { chatItemsSectionArea[it.key] = targetSection }
+      }
+    }
+    chatItemsSectionArea.putAll(itemsToAdd.associate { it.id to sectionArea })
```

Comment:
Waaay too much loops. Can this be done in one loop?

- for
- filter
- forEach
- associate

Reply:
for every item you do two loops + top level for...